tests/unit/lifecycle.ts

```diff
@@ -0,0 +1,261 @@
+import * as registerSuite from 'intern!object';
+import * as assert from 'intern/chai!assert';
+import { WidgetBase } from '../../src/WidgetBase';
+import { v } from '../../src/d';
+import { ProjectorMixin } from '../../src/mixins/Projector';
+import { WidgetProperties, HNode } from '../../src/interfaces';
+import { waitFor } from './waitFor';
+
+class TesterWidget extends WidgetBase<WidgetProperties> {
+	public lifeCycleCreated: Array<{key: string, element: Element}> = [];
+	public lifeCycleUpdated: Array<{key: string, element: Element}> = [];
+
+	protected vnode: HNode;
+	public modify: boolean = false;
+
+	protected onElementCreated(element: Element, key: string): void {
+		this.lifeCycleCreated.push({
+			key: key,
+			element: element
+		});
+	}
+
+	protected onElementUpdated(element: Element, key: string): void {
+		this.lifeCycleUpdated.push({
+			key: key,
+			element: element
+		});
+	}
+}
+
+/**
+ * Test widget that records the calls to the lifecycle methods.
+ */
+class WidgetA extends TesterWidget {
+
+	public render(): HNode {
+		let vnode = this.vnode;
+
+		if (vnode) {
+			if (this.modify) {
+				let child = <HNode> vnode.children[0]!;
+				child = <HNode> child.children[0]!;
+				child.children!.push(v('span', {key: 'addition', id: 'addition'}, [' Modified!']));
+			}
+		} else {
+			this.vnode = vnode = v('div', {
+				key: 'div1'
+			}, [
+				v('div', {}, [
+					v('div', {
+						key: 'div2'
+					}, ['This is a test.'])
+				])
+			]);
+		}
+		return vnode;
+	}
+}
+
+/**
+ * Test widget that adds afterCreate and afterUpdate callbacks to each vnode.
+ */
+let afterCreateCounter = 0;
+let afterUpdateCounter = 0;
+function incrementCreateCounter(): void {
+	afterCreateCounter++;
+}
+
+function incrementUpdateCounter(): void {
+	afterUpdateCounter++;
+}
+class WidgetB extends TesterWidget {
+
+	public render(): HNode {
+		let vnode = this.vnode;
+
+		if (vnode) {
+			if (this.modify) {
+				vnode.children.push(v('div', {
+					key: 'addition',
+					id: 'addition',
+					afterCreate: incrementCreateCounter.bind(this),
```

Comment:
Yeah I see. It would be nice if we could narrow the interface, but the way we currently have it makes that kind of tricky (because you still need it when you attach the afterCreate and afterUpdate). Happy to leave this for now